wpep0002.md

```diff
@@ -0,0 +1,58 @@
+| Key | Value |
+| ---: | :--- |
+| **Title** | Improve distributed capabilities |
+| **Number** | 0002 |
+| **Date** | 2019-10-21 |
+| **Authors** | Frederic Lepied <flepied@wazo.io> |
+| **Status** | proposed |
+
+## Abstract
+
+Redesign the way services find each other and their configurations.
+
+## Motivation
+
+To be able to cope with container deployements like kubernetes, swarm
+or docker-compose as well as distributed deployments on vm or bare
+metal, the startup sequences of the services need to be more dynamic
+without anything being done at install time.
+
+## Proposition
+
+Use Consul to discover services and settings.
+
+While the needed services are not started (nothing in Consul), the
+service shound wait for them to start and publish their informations.
+
+If a service needs a database, it should manage it at startup: create
+it if it doesn't exist and do the modifications if a database schema
+change is needed.
+
+The setup of the services themselves like RabbitMQ in cluster mode,
+load balancing database is out of scope of this proposition. What is
+exposed by this proposition is only the URI of the service not the way
+it has been deployed.
+
+To summarize, nothing should be done by the packages or the Ansible
+installer at install time that is deployment specific. For example, at
+install time, we must not generate SSL certs, create or manipulate
+databases or set any deployment specific parameter.
+
+For non Consul aware services like Asterisk, Kamailio, RTPEngine,
+PostgreSQL or RabbitMQ, we have 2 options. First one is to create a
+wrapper which is interfacing with Consul to publish the state and get
+the needed parameters. The second option is to write a plugin for the
+projects that support it like Asterisk or Kamailio to interface with
+Consul.
```

Comment:
RabbitMQ is Consul aware.

One interesting feature which hasn't been mentioned is the Consul DNS. You can resolve for example swarm.service.development.wazo. Or i-12345efg78.node.service.development.wazo.

A third option, for metal & VM at least is to run a consul-template service that'll write configuration files on Consul update, then run a command or script, most likely to reload the service.

A full example of this usage is our future reverse proxies. HAProxy backend are written by consul-template. Which trigger haproxy service reload. So I only need to add published in the consul service tags, plus a metadata "URL".

Reply:
We have discussed the DNS option but we wanted a solution that will work the same in containers and in VM/bare metal and that's why we didn't go this route.

We didn't really discussed the consult-templates option. I don't know if this is a good idea in containerized services.